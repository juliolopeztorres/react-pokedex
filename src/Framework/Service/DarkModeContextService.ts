import { createContext } from 'react'

const DarkModeContextService = createContext<[boolean, (status: boolean) => void]>([false, () => {}])

export default DarkModeContextService
