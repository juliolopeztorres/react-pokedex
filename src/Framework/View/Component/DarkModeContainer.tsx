import React, { useContext, useMemo, useState } from 'react'
import { hot } from "react-hot-loader/root";
import ServiceContainerContextService from "../../Service/ServiceContainerContextService";
import DarkModeContextService from "../../Service/DarkModeContextService";

const DarkModeContainer: (props: {children: React.ReactNode}) => React.ReactElement = ({children}) => {
  const serviceContainer = useContext(ServiceContainerContextService)

  const localStorage = useMemo<Storage>(
    () => serviceContainer.getService('LocalStorage'),
    []
  )

  const [darkMode, setDarkMode] = useState<boolean>('true' === localStorage.getItem('darkMode'))

  const onDarkModeChanged = (status : boolean) => {
    setDarkMode(status)
    localStorage.setItem('darkMode', status.toString())
  }

  return <div className={`${darkMode ? 'dark' : ''}`}>
    <div className="antialiased transition-colors duration-500 ease-in-out flex flex-col min-h-screen dark:bg-slate-800 dark:text-gray-200">
      <DarkModeContextService.Provider value={[darkMode, onDarkModeChanged]}>
        {children}
      </DarkModeContextService.Provider>
    </div>
  </div>
}

export default hot(DarkModeContainer)
