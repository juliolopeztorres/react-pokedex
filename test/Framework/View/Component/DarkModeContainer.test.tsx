import React, { useContext, useEffect } from 'react'
import { createRoot, Root } from "react-dom/client"
import { act } from "@testing-library/react";
import { clearTimeout } from "timers";
import DarkModeContextService from "../../../../src/Framework/Service/DarkModeContextService";
import DarkModeContainer from "../../../../src/Framework/View/Component/DarkModeContainer";

let root: Root
let container: HTMLElement
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);

  root = createRoot(container)
  jest.useFakeTimers()
});

afterEach(() => {
  // cleanup on exiting
  act(() => root.unmount())

  container.remove();
  jest.useRealTimers()
});

it('should render', () => {
  const Children = () => {
    const [darkMode, setDarkMode] = useContext(DarkModeContextService)

    useEffect(() => {
      const tid = setTimeout(() => {
        setDarkMode(true)
      }, 1000)

      return () => {
        clearTimeout(tid)
      }
    }, [])

    return <div>
      <span>Children to be drawn. Dark mode is {darkMode.toString()}</span>
    </div>
  }

  act(() => {
    root.render(<DarkModeContainer><Children/></DarkModeContainer>)
  })

  expect(container).toMatchSnapshot()

  act(() => {
    jest.advanceTimersByTime(1200)
  })

  expect(container).toMatchSnapshot()
});


it('should render using localStorage saved value', () => {
  localStorage.setItem('darkMode', 'true')

  const Children = () => {
    const [darkMode, setDarkMode] = useContext(DarkModeContextService)

    useEffect(() => {
      const tid = setTimeout(() => {
        setDarkMode(false)
      }, 1000)

      return () => {
        clearTimeout(tid)
      }
    }, [])

    return <div>
      <span>Children to be drawn. Dark mode is {darkMode.toString()}</span>
    </div>
  }

  act(() => {
    root.render(<DarkModeContainer><Children/></DarkModeContainer>)
  })

  expect(container).toMatchSnapshot()

  act(() => {
    jest.advanceTimersByTime(1200)
  })

  expect(container).toMatchSnapshot()
});
