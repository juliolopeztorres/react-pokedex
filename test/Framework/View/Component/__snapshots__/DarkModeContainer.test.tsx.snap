// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="antialiased transition-colors duration-500 ease-in-out flex flex-col min-h-screen dark:bg-slate-800 dark:text-gray-200"
    >
      <div>
        <span>
          Children to be drawn. Dark mode is 
          false
        </span>
      </div>
    </div>
  </div>
</div>
`

exports[`should render 2`] = `
<div>
  <div
    class="dark"
  >
    <div
      class="antialiased transition-colors duration-500 ease-in-out flex flex-col min-h-screen dark:bg-slate-800 dark:text-gray-200"
    >
      <div>
        <span>
          Children to be drawn. Dark mode is 
          true
        </span>
      </div>
    </div>
  </div>
</div>
`

exports[`should render using localStorage saved value 1`] = `
<div>
  <div
    class="dark"
  >
    <div
      class="antialiased transition-colors duration-500 ease-in-out flex flex-col min-h-screen dark:bg-slate-800 dark:text-gray-200"
    >
      <div>
        <span>
          Children to be drawn. Dark mode is 
          true
        </span>
      </div>
    </div>
  </div>
</div>
`

exports[`should render using localStorage saved value 2`] = `
<div>
  <div
    class=""
  >
    <div
      class="antialiased transition-colors duration-500 ease-in-out flex flex-col min-h-screen dark:bg-slate-800 dark:text-gray-200"
    >
      <div>
        <span>
          Children to be drawn. Dark mode is 
          false
        </span>
      </div>
    </div>
  </div>
</div>
`
